### PR TITLE
feat(snowflake): initial commit for snowflake backend

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds a snowflake backend to ibis. There are numerous issues to work out still, but many simple queries work.

### CODE REVIEW

File: ibis/backends/base/sql/alchemy/datatypes.py [hunk 1]
```diff
@@ -367,6 +368,58 @@ def sa_struct(dialect, satype, nullable=True):
...
+        TIMESTAMP_TZ,
+        VARIANT,
+    )
+    from snowflake.sqlalchemy.snowdialect import SnowflakeDialect
+except ImportError:
+    pass
+else:
+
+    @dt.dtype.register(SnowflakeDialect, (TIMESTAMP_LTZ, TIMESTAMP_TZ))
+    def sa_sf_timestamp_ltz(dialect, satype, nullable=True):
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/datatypes.py:386]: Could we place these into the snowflake backend?

File: ibis/backends/base/sql/__init__.py [hunk 2]
```diff
@@ -50,7 +51,12 @@ def _from_url(self, url: str) -> BaseBackend:
         kwargs = {
             name: value
             for name in ("host", "port", "database", "password")
-            if (value := getattr(url, name, None))
+            if (
+                value := (
+                    getattr(url, name, None)
+                    or os.environ.get(f"{self.name.upper()}_{name.upper()}")
```
Comment (kszucs) [inline ibis/backends/base/sql/__init__.py:57]: It's a bit hard to read.

**Comment by kszucs** [comment]
@cpcloud how can I test this locally?

**Comment by cpcloud** [comment]
@kszucs

1. Register for a snowflake accoutn
2. `export SNOWFLAKE_URL='snowflake://$user:$password@$locator.$region.$cloud/ibis_testing/ibis_testing'`
3. `pytest -m snowflake`

The `$locator` can be hard to find. It's **NOT** the string you see with 
image

when you loging.

It's the value under the `LOCATOR` column in the `Admin -> Accounts` panel.

Let me know if something isn't clear or isn't working.

**Comment by kszucs** [comment]
That url is not straightforward, I had to dig region cloud values from the documentation. 
Apparently I also need to configure a warehouse and append that warehouse's name to the URL as `?warehouse=<name>`

Still I'm getting `sqlalchemy.exc.InterfaceError: (snowflake.connector.errors.InterfaceError) 252005: Failed to convert current row, cause: [Snowflake Exception] unknown arrow internal data type(51669416) for FIXED data`

**Comment by cpcloud** [comment]
> That url is not straightforward, I had to dig region cloud values from the documentation. 
> Apparently I also need to configure a warehouse and append that warehouse's name to the URL as `?warehouse=<name>`

It's not very straightforward, I agree. That's an issue to take up over at the sqlalchemy dialect for snowflake.

> 
> Still I'm getting `sqlalchemy.exc.InterfaceError: (snowflake.connector.errors.InterfaceError) 252005: Failed to convert current row, cause: [Snowflake Exception] unknown arrow internal data type(51669416) for FIXED data`

Their arrow integration has all sorts of problems. I would try without arrow (you may have to uninstall pyarrow) to see if that works.

**Comment by kszucs** [comment]
With a downgraded pyarrow==6 it works.

**Comment by kszucs** [comment]
I have the following test errors locally:

```
FAILED ibis/backends/tests/test_numeric.py::test_sa_default_numeric_precision_and_scale[snowflake-default_precisions0-default_scales0] - sqlalchemy.exc.ProgrammingError: (snowflake.connector.errors.ProgrammingError) 002002 (42710): SQL compilati...
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_column_group_by[snowflake-day_of_week_full_name] - AssertionError: DataFrame.iloc[:, 1] (column name="day_of_week_result") are different
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-sunday] - AssertionError: assert 'day' == 'sunday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-monday] - AssertionError: assert 'day' == 'monday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-thursday] - AssertionError: assert 'day' == 'thursday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_column[snowflake] - AssertionError: Series are different
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-friday] - AssertionError: assert 'day' == 'friday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-tuesday] - AssertionError: assert 'day' == 'tuesday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-wednesday] - AssertionError: assert 'day' == 'wednesday'
FAILED ibis/backends/tests/test_temporal.py::test_day_of_week_scalar[snowflake-saturday] - AssertionError: assert 'day' == 'saturday'
```

**Comment by cpcloud** [comment]
Hm ok. I will look into that then.

File: ibis/backends/base/sql/alchemy/datatypes.py (hunk 1 above)
Comment (cpcloud) [reply ibis/backends/base/sql/alchemy/datatypes.py:386]: I think so, trying now.

File: ibis/backends/snowflake/datatypes.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        | spaceless_string("object").result(Map(string, json))
+        | spaceless_string("array").result(Array(json))
+        | spaceless_string("variant").result(json)
+        | varchar
+        | decimal
+    )
+    return ty.parse(text)
+
+
+@dt.dtype.register(SnowflakeDialect, (TIMESTAMP_LTZ, TIMESTAMP_TZ))
```
Comment (kszucs) [inline ibis/backends/snowflake/datatypes.py:140]: We need to clean up the datatype conversions for the sqlalchemy backends.

**Review by kszucs — APPROVED** [review]
LGTM

I'll go ahead and merge this. We can address upcoming issues in follow-up PRs.

**Pull request merged**